File: core/src/edu/gvsu/cis/spacejourney/level/one/LevelOne.java
package edu.gvsu.cis.spacejourney.level.one;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.audio.Music;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.scenes.scene2d.Stage;
import edu.gvsu.cis.spacejourney.SpaceJourney;
import edu.gvsu.cis.spacejourney.entity.PlayerSpaceship;
import edu.gvsu.cis.spacejourney.entity.collectible.Collectible;
import edu.gvsu.cis.spacejourney.entity.collectible.TestCollectible;
import edu.gvsu.cis.spacejourney.entity.enemy.EvilSpaceship;
import edu.gvsu.cis.spacejourney.entity.movement.LinearMovement;
import edu.gvsu.cis.spacejourney.input.PlayerInputListener;
import edu.gvsu.cis.spacejourney.level.Level;
import edu.gvsu.cis.spacejourney.level.choreography.LevelChoreographer;
import edu.gvsu.cis.spacejourney.level.choreography.events.EntitySpawnEvent;
import edu.gvsu.cis.spacejourney.screens.backgrounds.ParallaxBackground;
import edu.gvsu.cis.spacejourney.screens.hud.DefaultOverlay;
import edu.gvsu.cis.spacejourney.util.ZIndex;

public class LevelOne extends Level {

  private static final int LEVEL_TIME = 200;

  private ParallaxBackground background;
  private PlayerSpaceship player;
  private PlayerInputListener inputListener;
  private Collectible testCollectible;

  private LevelChoreographer choreographer;
  private DefaultOverlay defaultHud;

  /**
   * Default constructor for the first level.
   */
  public LevelOne() {
    setMusic(SpaceJourney.Companion.getAssetManager().get(
        "Space Background Music.mp3", Music.class));
  }

  /**
   * Method that initializes logic specific to the first level.
   * @param stage for the current stage of the game.
   * @param world for the current world of the game.
   */
  @Override
  public void init(Stage stage, World world) {
    super.init(stage, world);

    background = new ParallaxBackground();
    background.setZIndex(ZIndex.BACKGROUND);
    stage.addActor(background);

    player = new PlayerSpaceship(stage);
    player.setPosition(1.5f, 0.0f);
    player.setWidth(50.0f);
    player.setHeight(50.0f);
    player.createBody(world);
    stage.addActor(player);
    setPlayer(player);

    defaultHud = new DefaultOverlay();
    setHud(defaultHud);

    inputListener = new PlayerInputListener(player);
    Gdx.input.setInputProcessor(inputListener);

    testCollectible = new TestCollectible(stage);
    testCollectible.createBody(world);
    stage.addActor(testCollectible);

    choreographer = new LevelChoreographer(stage, world);

    for (int i = 0; i < LEVEL_TIME; i++) {
      EvilSpaceship basicEnemy = new EvilSpaceship(stage);
      basicEnemy.setWidth(35.0f);
      basicEnemy.setHeight(35.0f);
      basicEnemy.setMovementPattern(new LinearMovement(new Vector2(0f, -25f)));
      choreographer.schedule(1.0f + i, new EntitySpawnEvent(basicEnemy));
    }

  }

  /**
   * Method that updates the logic each time the render() method is called.
   * @param delta float given during a screen's render() method.
   */
  @Override
  public void update(float delta) {
    inputListener.poll(delta);
    choreographer.update(delta);
    defaultHud.poll();

    if (choreographer.isEmpty()) {
      EntitySpawnEvent event = (EntitySpawnEvent) choreographer.getLastEvent().getEvent();
      if (event.getEntity().belowScreen()) {
        // TODO - Implement level completed... or something...
    	  System.out.println("FINISHED");
      }
    }

  }

  /**
   * Disposes of any disposable objects, such as music, textures, or entities.
   */
  @Override
  public void dispose() {
    player.dispose();
    testCollectible.dispose();
    getMusic().stop();
    getMusic().dispose();
  }
}

File: core/src/edu/gvsu/cis/spacejourney/level/one/LevelOne.kt
package edu.gvsu.cis.spacejourney.level.one

import com.badlogic.ashley.core.Engine
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.audio.Music
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.math.Vector2
import edu.gvsu.cis.spacejourney.SpaceJourney
import edu.gvsu.cis.spacejourney.component.Parallax
import edu.gvsu.cis.spacejourney.component.StaticSprite
import edu.gvsu.cis.spacejourney.component.Transform
import edu.gvsu.cis.spacejourney.level.Level
import edu.gvsu.cis.spacejourney.level.choreography.events.EnemySpawnEvent
import edu.gvsu.cis.spacejourney.level.choreography.LevelChoreographer
import edu.gvsu.cis.spacejourney.managers.GameDataManager
import edu.gvsu.cis.spacejourney.screens.hud.DefaultOverlay
import edu.gvsu.cis.spacejourney.system.PlayerControllerSystem
import edu.gvsu.cis.spacejourney.util.ZIndex
import ktx.ashley.add
import ktx.ashley.entity
import ktx.log.debug

//private DefaultOverlay defaultHud;

class LevelOne : Level() {

    private var choreographer: LevelChoreographer? = null

    override fun init(engine: Engine) {
        super.init(engine)

        this.hud = DefaultOverlay()
        this.music = SpaceJourney.assetManager.get("Space Background Music.mp3", Music::class.java)

        engine.add {
            entity {
                with<Transform> {
                    position = Vector2(0f, 0f)
                }
                with<StaticSprite> {
                    zindex = ZIndex.PARALLAX_BACKGROUND_LAYER1
                    texture = SpaceJourney.assetManager.get("parallax_medium_star_layer.png", Texture::class.java)
                    repeating = true
                    size = Vector2(Gdx.graphics.width.toFloat(), Gdx.graphics.height.toFloat())
                    scale = 1.0f
                }
                with<Parallax> {
                    speed = 0.25f
                }
            }
            entity {
                with<Transform> {
                    position = Vector2(0f, 0f)
                }
                with<StaticSprite> {
                    zindex = ZIndex.PARALLAX_BACKGROUND_LAYER2
                    texture = SpaceJourney.assetManager.get("parallax_planet_layer.png", Texture::class.java)
                    repeating = true
                    size = Vector2(Gdx.graphics.width.toFloat(), Gdx.graphics.height.toFloat())
                    scale = 1.0f
                }
                with<Parallax> {
                    speed = 0.1f
                }
            }
            entity {
                with<Transform> {
                    position = Vector2(0f, 0f)
                }
                with<StaticSprite> {
                    zindex = ZIndex.PARALLAX_BACKGROUND_LAYER3
                    texture = SpaceJourney.assetManager.get("parallax_small_star_layer.png", Texture::class.java)
                    repeating = true
                    size = Vector2(Gdx.graphics.width.toFloat(), Gdx.graphics.height.toFloat())
                    scale = 1.0f
                }
                with<Parallax> {
                    speed = 0.075f
                }
            }
        }

        choreographer = LevelChoreographer(engine)

        for (i in 0..49) {
            choreographer!!.schedule(1.0f + i * 0.2f, EnemySpawnEvent())
        }
    }

    override fun update(delta: Float) {
        super.update(delta)

        choreographer!!.update(delta)
        hud?.poll(engine!!)

        if (choreographer!!.isEmpty()) {
            val lastEvent = choreographer!!.getLastEvent()?.event as? EnemySpawnEvent
            val inputEnabled = engine?.getSystem(PlayerControllerSystem::class.java)?.inputEnabled
            if (!engine!!.entities.contains(lastEvent?.enemyEntity) && inputEnabled!!) {
                // TODO - Implement level end transition!
                engine?.getSystem(PlayerControllerSystem::class.java)?.inputEnabled = false
                debug { "Last entity removed. Starting level end sequence." }

            }
        }
    }

    override fun dispose() {
        music?.stop()
        music?.dispose()
    }
}
